File: Implementations/RSA.r
# import openssl:
library(openssl)
# C equivalent: (since most algorithms have reference implementations based on the C variant)
# #include <openssl/rsa.h>

# The main function we are concerned with is rsa_keygen(), which generates a random private key of specified bits.
private_key <- rsa_keygen(bits = 1024)
# Now that we got our 1024-bit private key, we can extract the public key with the same number of bits:
public_key <- private_key$public_key

# Encryption & Decryption can be followed with rsa_encrypt() and rsa_decrypt() functions:
message <- charToRaw("Alias{Anirban166}: Raven")
ciphertext <- rsa_encrypt(message, public_key)
rawToChar(rsa_decrypt(ciphertext, private_key)) # prints out the message

# To see whats underneath, (such as the numbers use for the key) we can unfold the private key:
private_key$data
# It includes the 1024-bit prime numbers included for key generation (p,q), their product n or the Euler's totient and the private & public exponents (e,d).
# On the other hand, the public key only contains the public & private exponents (e,d) and the product of the primes (n), making it impossible to decrypt the message or find the two numbers (p,q).

# For encrypting the plaintext message into ciphertext, we need to convert it into an integer:
plain.text <- bignum(charToRaw("Alias{Anirban166}: Raven"))
# Extract the Euler's totient and the exponent e (public) from the public key:
n <- public_key$data$n
e <- public_key$data$e
# Calculate the ciphertext:
cipher.text <- (m ^ e) %% n
# To obtain the exponent d (private), we need to use the private key:
d <- private_key$data$d
# Since the value of ciphertext raised to the power of d is extremely large, we will need to use bignum_mod_exp() function in order to decrypt the ciphertext:
decrypted.message <- bignum_mod_exp(cipher.text, d, n)
# Again, we need to convert the result into characters to obtain the string we sent as a message:
rawToChar(decrypted.message) # prints out the message

File: Implementations/RSA.cpp
#include <iostream>
#include <cmath>
#include <cstring>
#include <cstdlib>
using l = long long int;
#define S 100

// Globally passed variables:
l p, q,      // primes
  n,        // n = pq
  t,       // toutient
  flag,   // 1/0
  e[S],  // public exponent e
  d[S], // private exponent d
  temp[S], m[S], en[S];

char msg[S];

// Functions:
bool primecheck(l); // Check if(prime)
void ce();         // Calculate exponent e
l cd(l);          // Calculate exponent d (return value)
void encrypt();  // Encrypt plaintext
void decrypt(); // Decrypt ciphertext

bool primecheck(l var) 
{
	int i, j = sqrt(var);
	for (i = 2; i <= j; ++i)
	if (var % i == 0)
			return false;
	return true;
}

void ce()
{
     int k = 0;
     for (int i = 2; i < t; i++)
     {
	 if (t % i == 0) continue;
	 flag = primecheck(i);
	 if (flag == 1 && i != p && i != q)
	 {
	     e[k] = i;
	     flag = cd(e[k]);
	     if (flag > 0)
	     {
		d[k] = flag;
		k++;
	     }
	     if (k == 99) // (S-1)
	     break;
	 }
     }
}

l cd(l x) 
{
     long int k = 1;
     while (true)
     {
	 k += t;
	 if (k % x == 0) return (k / x);
     }
}

void encrypt()
{
	l pt, ct, key = e[0], k, len;
	int i = 0;
	len = strlen(msg);
	while (i != len)
	{
		pt = m[i];
		pt = pt - 96;
		k = 1;
		for (int j = 0; j < key; j++)
		{
			k *= pt;
			k %= n;
		}
		temp[i] = k;
		ct = k + 96;
		en[i] = ct;
		i++;
	}
	en[i] = -1;
	std::cout << "Encrypted message:\n";
	for (i = 0; en[i] != -1; i++)
		printf("%c", en[i]);
}

void decrypt() 
{
	l pt, ct, key = d[0], k;
	int i = 0;
	while (en[i] != -1)
	{
		ct = temp[i];
		k = 1;
		for (int j = 0; j < key; j++)
		{
			k *= ct;
			k %= n;
		}
		pt = k + 96;
		m[i] = pt;
		i++;
	}
	m[i] = -1;
	std::cout << "Decrypted message:\n";
	for (int i = 0; m[i] != -1; i++)
		printf("%c", m[i]);
}

int main()
{   
    do 
    {
	   std::cout << "Enter two primes (p,q for Euler's totient to be taken as (p-1)(q-1)):\n";
	   std::cin >> p >> q;
	   if(primecheck(p) && primecheck(q)) break;
	   else std::cout << "Wrong input; Please enter prime numbers.\n";
    } while(1);    

	std::cout << "Enter message for encryption:\n";
	fflush(stdin);
	std::cin >> msg;
	for (int i = 0; msg[i] != '\0'; i++)
		m[i] = msg[i];
		
	n = p * q;
	t = (p - 1) * (q - 1);
	ce();
	encrypt();
	decrypt();
	return 0;
}
